File: tabulation/javascript/bestSum.js
function bestSum(targetSum, numbers) {
    const table = Array(targetSum + 1).fill(null)
    table[0] = []

    for (let i = 0; i <= targetSum; i++) {
        if (table[i]) {
            for (num of numbers) {
                const candidateConbination = [...table[i], num]
                if (!table[i + num] || candidateConbination.length < table[i + num].length)
                    table[i + num] = candidateConbination
            }
        }
    }

    return table[targetSum]
}


console.log(bestSum(7, [2, 3]))
console.log(bestSum(57, [4, 6]))
console.log(bestSum(134, [4, 6]))

File: tabulation/kotlin/bestSum.kt
fun bestSum(targetSum: Int, numbers: Array<Int>): List<Int>? {
    val table: Array<List<Int>?> = Array(targetSum + 1) { null }
    table[0] = listOf()

    (0..targetSum).forEach {
        if (table[it] != null) {
            numbers.forEach { num ->
                val candidateCombination = table[it]!!.plus(num)
                if (it + num <= targetSum && (table[it + num] == null
                || candidateCombination.size < table[it + num]!!.size))
                    table[it + num] = candidateCombination
            }
        }
    }

    return table[targetSum]
}

fun main(): Unit {
    println(bestSum(7, arrayOf(2, 3)))
    println(bestSum(51, arrayOf(4, 7, 8)))
    println(bestSum(71, arrayOf(6, 9)))

}